solidity/contracts/isms/hook/PolygonZkevmIsm.sol

```diff
@@ -0,0 +1,205 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.8.0;
+
+/*@@@@@@@       @@@@@@@@@
+ @@@@@@@@@       @@@@@@@@@
+  @@@@@@@@@       @@@@@@@@@
+   @@@@@@@@@       @@@@@@@@@
+    @@@@@@@@@@@@@@@@@@@@@@@@@
+     @@@@@  HYPERLANE  @@@@@@@
+    @@@@@@@@@@@@@@@@@@@@@@@@@
+   @@@@@@@@@       @@@@@@@@@
+  @@@@@@@@@       @@@@@@@@@
+ @@@@@@@@@       @@@@@@@@@
+@@@@@@@@@       @@@@@@@@*/
+
+// ============ Internal Imports ============
+
+import {LibBit} from "../../libs/LibBit.sol";
+import {Message} from "../../libs/Message.sol";
+import {TypeCasts} from "../../libs/TypeCasts.sol";
+
+import {AbstractMessageIdAuthorizedIsm} from "../hook/AbstractMessageIdAuthorizedIsm.sol";
+
+import {IInterchainSecurityModule} from "../../interfaces/IInterchainSecurityModule.sol";
+import {IMailbox} from "../../interfaces/IMailbox.sol";
+import {ICcipReadIsm} from "../../interfaces/isms/ICcipReadIsm.sol";
+
+// ============ External Imports ============
+
+import {IPolygonZkEVMBridge} from "../../interfaces/polygonZkevm/IPolygonZkEVMBridge.sol";
+import {IBridgeMessageReceiver} from "../../interfaces/polygonZkevm/IBridgeMessageReceiver.sol";
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+
+/**
+ * @title PolygonZkevmIsm
+ * @notice Polygon zkEVM chain Ism that uses the Polygon zkEVM bridge to verify messages
+ */
+contract PolygonZkevmIsm is
+    ICcipReadIsm,
+    AbstractMessageIdAuthorizedIsm,
+    IBridgeMessageReceiver
+{
+    using Message for bytes;
+    using LibBit for uint256;
+    using TypeCasts for bytes32;
+    using Address for address payable;
+
+    IMailbox public mailbox;
+    string[] public offchainUrls;
+
+    // ============ Constants ============
+    IPolygonZkEVMBridge public immutable zkEvmBridge;
+    uint8 public constant override moduleType =
+        uint8(IInterchainSecurityModule.Types.CCIP_READ);
+    uint32 public immutable zkEvmBridgeDestinationNetId;
+
+    // ============ Constructor ============
+    constructor(
+        address _zkEvmBridge,
+        uint32 _zkEvmBridgeDestinationNetId,
+        address _mailbox,
+        string[] memory _offchainUrls
+    ) {
+        require(
+            Address.isContract(_zkEvmBridge),
+            "PolygonZkevmIsm: invalid ZkEVMBridge"
+        );
+        require(
+            Address.isContract(_mailbox),
+            "PolygonZkevmIsm: invalid Mailbox"
+        );
+        require(
+            _zkEvmBridgeDestinationNetId <= 1,
+            "PolygonZkevmIsm: invalid ZkEVMBridge destination network id"
+        );
+        zkEvmBridgeDestinationNetId = _zkEvmBridgeDestinationNetId;
+        zkEvmBridge = IPolygonZkEVMBridge(_zkEvmBridge);
+        mailbox = IMailbox(_mailbox);
+        offchainUrls = _offchainUrls;
+    }
+
+    /**
+     * @dev off-chain verification information for a given message.
+     * @param _message The message for which off-chain verification information is requested.
+     */
+    function getOffchainVerifyInfo(
```

Comment:
why do we need CCIP here, the relayer will be calling the verify function, no?

Reply:
maybe helpful to make a quick mermaid diagram.

Reply:
yeah, that's something that was needed to be discussed as bridge needs to generate a proof to claim. the ccip call was for the hyperlane relayer to get that proof using PolygonZKevm api (tho that does not follow the ccip spec.


https://github.com/curlypoint/hyperlane-zkevm-ccip